src/libraries/System.Console/src/System/Console.cs

```diff
@@ -33,21 +34,36 @@ public static class Console
         private static ConsoleCancelEventHandler? s_cancelCallbacks;
         private static ConsolePal.ControlCHandlerRegistrar? s_registrar;
 
-        internal static T EnsureInitialized<T>([NotNull] ref T? field, Func<T> initializer) where T : class =>
-            LazyInitializer.EnsureInitialized(ref field, ref InternalSyncObject, initializer);
+        internal static T EnsureInitialized<T>([NotNull] ref T? field, T value) where T : class
+            // Interlocked.CompareExchange ensures we don't overwrite a field that was set.
+            => Interlocked.CompareExchange(ref field, value, null) ?? value;
 
-        public static TextReader In => EnsureInitialized(ref s_in, () => ConsolePal.GetOrCreateReader());
+        internal static T EnsureInitializedDisposable<T>([NotNull] ref T? field, T value, IDisposable disposable) where T : class
+        {
+            if (Interlocked.CompareExchange(ref field, value, null) != null)
+            {
+                disposable.Dispose();
+            }
+
+            return field;
+        }
+
+        public static TextReader In => Volatile.Read(ref s_in) ?? ConsolePal.EnsureInitializedIn(ref s_in);
 
         public static Encoding InputEncoding
         {
             get
             {
-                return EnsureInitialized(ref s_inputEncoding, () => ConsolePal.InputEncoding);
+                return Volatile.Read(ref s_inputEncoding) ?? EnsureInitializedInputEncoding();
+
+                static Encoding EnsureInitializedInputEncoding()
+                    => EnsureInitialized(ref s_inputEncoding, ConsolePal.InputEncoding);
             }
             set
             {
                 CheckNonNull(value, nameof(value));
-                lock (InternalSyncObject)
+
+                lock (SyncObject)
```

Comment:
Is there still a race condition between one thread setting the `InputEncoding` property, and other thread initializing the `In` stream? I believe that the `In` stream can end up with a stale value of the encoding.

Reply:
I thought it is not an issue since we're first we're setting `s_in` to `null` after setting `s_encoding`?
```
                    Volatile.Write(ref s_inputEncoding, (Encoding)value.Clone());
                    Volatile.Write(ref s_in, null);
```

Reply:
Consider this program:

Thread 1:
```
    Console.InputEncoding = encoding1
    _ = Console.In;
```

Thread 2:
```
    Console.InputEncoding = encoding2
```

And this interleaving:

Thread 1: 
- Sets the encoding to encoding1
- Starts creating the TextReader from encoding1

Thread 2:
- Sets the encoding to encoding2
- Done.

Thread 3:
- Finishes initialization of the TextReader and sets s_in to point to it

Result: `In` is using encoding1, but `IntputEncoding` is set to encoding2